File: lattice_hash_demo.js
const { blake3 } = require('@noble/hashes/blake3');
const { bytesToHex } = require('@noble/hashes/utils');
const bs58 = require('bs58');

class LtHash {
  constructor() {
    this.hasher = blake3.create({});
  }

  init() {
    this.hasher = blake3.create({});
  }

  append(data) {
    this.hasher.update(data);
  }

  fini() {
    // Simulating 2048-byte XOF by creating multiple 32-byte chunks
    let fullOutput = [];
    let tempHash = this.hasher.digest(); // Get the initial digest (32 bytes)

    while (fullOutput.length < 2048) {
      // Rehash the digest to produce more chunks
      let chunk = blake3(tempHash);
      fullOutput.push(...chunk);
      tempHash = chunk; // Set the next input as the previous output
    }

    return Buffer.from(fullOutput.slice(0, 2048));
  }

  static add(a, b) {
    const result = Buffer.alloc(2048);
    for (let i = 0; i < 2048; i += 2) {
      const valA = a.readUInt16LE(i);
      const valB = b.readUInt16LE(i);
      result.writeUInt16LE((valA + valB) & 0xffff, i);
    }
    return result;
  }

  static sub(a, b) {
    const result = Buffer.alloc(2048);
    for (let i = 0; i < 2048; i += 2) {
      const valA = a.readUInt16LE(i);
      const valB = b.readUInt16LE(i);
      result.writeUInt16LE((valA - valB) & 0xffff, i);
    }
    return result;
  }

  static out(hash) {
    return bytesToHex(blake3(hash)).slice(0, 32); // Return 32-byte hex digest
  }
}

function generateRandomAccount() {
  return {
    lamports: Math.floor(Math.random() * 1e9), // Random lamports value
    data: Buffer.from(blake3(new Uint8Array(128))), // Randomized 128-byte data
    is_executable: Buffer.from([Math.random() < 0.5 ? 1 : 0]), // 1-byte boolean
    owner: Buffer.from(blake3(new Uint8Array(32))), // 32-byte owner
    pubkey: bs58.encode(Buffer.from(blake3(new Uint8Array(32)))), // base58 public key
  };
}

function computeLtHash(account) {
  if (account.lamports === 0) {
    return Buffer.alloc(2048, 0); // All zeros for zero lamports
  }

  const lthash = new LtHash();
  lthash.init();
  lthash.append(Buffer.from(account.lamports.toString()));
  lthash.append(account.data);
  lthash.append(account.is_executable);
  lthash.append(account.owner);
  lthash.append(Buffer.from(account.pubkey));
  return lthash.fini();
}

function main(numAccounts) {
  const accounts = Array.from({ length: numAccounts }, generateRandomAccount);

  console.log(`\n--- Lattice Hash Test for ${numAccounts} Accounts ---`);

  let combinedHash = Buffer.alloc(2048, 0); // Initial zeroed hash

  // Measure addition time
  console.time("Time for Adding All Accounts");
  accounts.forEach((account) => {
    const hash = computeLtHash(account);
    combinedHash = LtHash.add(combinedHash, hash);
  });
  console.timeEnd("Time for Adding All Accounts");

  const initialCombinedHashHex = LtHash.out(combinedHash);
  console.log(`Initial Combined Hash (first 16 chars): ${initialCombinedHashHex.slice(0, 16)}`);

  // Remove the last 50 accounts
  console.time("Time for Removing Last 50 Accounts");
  const last50Accounts = accounts.slice(-50);
  last50Accounts.forEach((account) => {
    const hash = computeLtHash(account);
    combinedHash = LtHash.sub(combinedHash, hash);
  });
  console.timeEnd("Time for Removing Last 50 Accounts");

  const afterRemovalHashHex = LtHash.out(combinedHash);
  console.log(`Hash after Removing Last 50 Accounts (first 16 chars): ${afterRemovalHashHex.slice(0, 16)}`);

  // Add the last 50 accounts back
  console.time("Time for Adding Last 50 Accounts Back");
  last50Accounts.forEach((account) => {
    const hash = computeLtHash(account);
    combinedHash = LtHash.add(combinedHash, hash);
  });
  console.timeEnd("Time for Adding Last 50 Accounts Back");

  const finalCombinedHashHex = LtHash.out(combinedHash);
  console.log(`Final Combined Hash (first 16 chars): ${finalCombinedHashHex.slice(0, 16)}`);

  console.log(`\nVerification: ${initialCombinedHashHex === finalCombinedHashHex ? "PASS ✅" : "FAIL ❌"}`);
}

const numAccounts = process.argv[2] ? parseInt(process.argv[2]) : 10000;
main(numAccounts);

File: lattice_hash_demo.py
import sys
import time
import random
import base58
import numpy as np
import multiprocessing as mp
from blake3 import blake3
from numba import njit, prange

class LtHash:
    def __init__(self):
        """ Initializes an instance of LtHash. """
        self.hasher = None

    def init(self):
        """ Initializes the blake3 hasher. """
        self.hasher = blake3()

    def append(self, data: bytes):
        """ Appends data to the hash. """
        self.hasher.update(data)

    def fini(self):
        """ Finalizes the hash by extending the output to 2048 bytes. """
        temp_hash = self.hasher.digest()
        full_output = bytearray()

        while len(full_output) < 2048:
            temp_hash = blake3(temp_hash).digest()
            full_output.extend(temp_hash)

        return bytes(full_output[:2048])

    @staticmethod
    @njit(parallel=True, fastmath=True)
    def add(a: np.ndarray, b: np.ndarray) -> np.ndarray:
        """ Element-wise 16-bit addition with modulo 65536. """
        result = a.astype(np.uint32) + b.astype(np.uint32)
        return (result & np.uint16(0xFFFF)).astype(np.uint16)

    @staticmethod
    @njit(parallel=True, fastmath=True)
    def sub(a: np.ndarray, b: np.ndarray) -> np.ndarray:
        """ Element-wise 16-bit subtraction with modulo 65536. """
        result = a.astype(np.uint32) - b.astype(np.uint32)
        return (result & np.uint16(0xFFFF)).astype(np.uint16)

    @staticmethod
    def out(hash_bytes: bytes) -> str:
        """ Returns the first 32 characters of the hex-encoded blake3 digest. """
        return blake3(hash_bytes).digest().hex()[:32]


def generate_random_account():
    """ Generates a random account with blake3 hashed fields. """
    return {
        'lamports': random.randint(0, int(1e9)),
        'data': blake3(bytes(128)).digest(),
        'is_executable': bytes([random.randint(0, 1)]),
        'owner': blake3(bytes(32)).digest(),
        'pubkey': base58.b58encode(blake3(bytes(32)).digest()).decode('utf-8'),
    }


def compute_lt_hash(account):
    """ Computes the lattice hash for a given account. """
    if account['lamports'] == 0:
        return np.zeros(1024, dtype=np.uint16)

    lthash = LtHash()
    lthash.init()

    lthash.append(str(account['lamports']).encode('utf-8'))
    lthash.append(account['data'])
    lthash.append(account['is_executable'])
    lthash.append(account['owner'])
    lthash.append(account['pubkey'].encode('utf-8'))

    return np.frombuffer(lthash.fini(), dtype=np.uint16)


def process_accounts_parallel(accounts, combined_hash, add=True):
    """ Processes accounts in parallel, adding or subtracting their hashes. """
    num_workers = mp.cpu_count()
    pool = mp.Pool(num_workers)

    hashes = pool.map(compute_lt_hash, accounts)
    pool.close()
    pool.join()

    for h in hashes:
        if add:
            combined_hash[:] = LtHash.add(combined_hash, h)  # Modify in place
        else:
            combined_hash[:] = LtHash.sub(combined_hash, h)  # Modify in place

    return combined_hash


def main(num_accounts: int):
    """ Main function to generate accounts and compute the lattice hash. """
    accounts = [generate_random_account() for _ in range(num_accounts)]
    print(f"\n--- Lattice Hash Test for {num_accounts} Accounts ---")

    combined_hash = np.zeros(1024, dtype=np.uint16)

    # Step 1: Compute initial hash
    start = time.time()
    combined_hash = process_accounts_parallel(accounts, combined_hash, add=True)
    end = time.time()
    print(f"Time for Adding All Accounts: {end - start:.4f} seconds")

    initial_hash_hex = LtHash.out(combined_hash.tobytes())
    print(f"Initial Combined Hash (first 16 chars): {initial_hash_hex[:16]}")

    # Step 2: Remove the last 50 accounts
    last_50 = accounts[-50:] if len(accounts) >= 50 else accounts
    start = time.time()
    combined_hash = process_accounts_parallel(last_50, combined_hash, add=False)
    end = time.time()
    print(f"Time for Removing Last 50 Accounts: {end - start:.4f} seconds")

    after_removal_hash_hex = LtHash.out(combined_hash.tobytes())
    print(f"Hash after Removing Last 50 Accounts (first 16 chars): {after_removal_hash_hex[:16]}")

    # Step 3: Add back last 50 accounts
    start = time.time()
    combined_hash = process_accounts_parallel(last_50, combined_hash, add=True)
    end = time.time()
    print(f"Time for Adding Last 50 Accounts Back: {end - start:.4f} seconds")

    final_hash_hex = LtHash.out(combined_hash.tobytes())
    print(f"Final Combined Hash (first 16 chars): {final_hash_hex[:16]}")

    # Final verification
    print(f"\nVerification: {'PASS ✅' if (initial_hash_hex == final_hash_hex) else 'FAIL ❌'}")


if __name__ == "__main__":
    num_accounts = int(sys.argv[1]) if len(sys.argv) > 1 else 10000
    main(num_accounts)
